src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -299,33 +334,208 @@ private BoundExpression RewriteStringConcatenationOneExpr(BoundExpression lowere
+
+            if ((firstIsChar || secondIsChar || thirdIsChar || firstType.IsReadOnlySpanChar() || secondType.IsReadOnlySpanChar() || thirdType.IsReadOnlySpanChar()) &&
+                TryGetWellKnownTypeMember(syntax, WellKnownMember.System_String__Concat_ReadOnlySpanReadOnlySpanReadOnlySpan, out MethodSymbol stringConcatSpans, isOptional: true) &&
+                TryGetWellKnownTypeMember(syntax, WellKnownMember.System_String__op_Implicit_ToReadOnlySpanOfChar, out MethodSymbol stringImplicitConversionToReadOnlySpan, isOptional: true) &&
+                TryGetWellKnownTypeMember(syntax, WellKnownMember.System_ReadOnlySpan_T__ctor_Reference, out MethodSymbol readOnlySpanWrappingCtorGeneric, isOptional: true))
+            {
+                // One of args is a char or a ReadOnlySpan<char> (e.g. from previous rewrite). We can use span-based concatenation, which takes a bit more IL, but avoids allocation of a string from ToString call in case of char.
+                // And since implicit conversion from string to span is a JIT intrinsic, the resulting code ends up being faster than the one generated from the "naive" case with ToString call
+                var readOnlySpanOfChar = readOnlySpanWrappingCtorGeneric.ContainingType.Construct(_compilation.GetSpecialType(SpecialType.System_Char));
+                var readOnlySpanCtorRefParamChar = readOnlySpanWrappingCtorGeneric.AsMember(readOnlySpanOfChar);
+
+                var locals = ArrayBuilder<LocalSymbol>.GetInstance(capacity: previousLocals.Count + 1);
+                locals.AddRange(previousLocals);
+
+                var wrappedFirst = WrapIntoReadOnlySpanOfCharIfNecessary(loweredFirst, stringImplicitConversionToReadOnlySpan, readOnlySpanCtorRefParamChar, out var firstTempLocal);
+                var wrappedSecond = WrapIntoReadOnlySpanOfCharIfNecessary(loweredSecond, stringImplicitConversionToReadOnlySpan, readOnlySpanCtorRefParamChar, out var secondTempLocal);
+                var wrappedThird = WrapIntoReadOnlySpanOfCharIfNecessary(loweredThird, stringImplicitConversionToReadOnlySpan, readOnlySpanCtorRefParamChar, out var thirdTempLocal);
+
+                locals.AddIfNotNull(firstTempLocal);
+                locals.AddIfNotNull(secondTempLocal);
+                locals.AddIfNotNull(thirdTempLocal);
+
+                var concatCall = BoundCall.Synthesized(syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, stringConcatSpans, ImmutableArray.Create(wrappedFirst, wrappedSecond, wrappedThird));
+
+                return new BoundSequence(
+                    syntax,
+                    locals.ToImmutableAndFree(),
+                    ImmutableArray<BoundExpression>.Empty,
+                    concatCall,
+                    concatCall.Type);
+            }
+
+            Debug.Assert(!firstType.IsReadOnlySpanChar() && !secondType.IsReadOnlySpanChar() && !thirdType.IsReadOnlySpanChar());
```

Comment:
>Debug.Assert(!firstType.IsReadOnlySpanChar() && !secondType.IsReadOnlySpanChar() && !thirdType.IsReadOnlySpanChar()); [](http://example.com/codeflow?start=12&length=117)

This assumption looks suspicious to me. I saw your response to Fred's comment, but I remain skeptical because we require presence of more helpers in the previous ```if```. At the very least we can benefit from a comment containing complete proof of the fact that we are asserting.

Reply:
> This assumption looks suspicious to me. I saw your response to Fred's comment, but I remain skeptical ...

Here is a unit-test for which this assert fails. Once it is disabled, PEVerify fails and execution fails with "System.InvalidProgramException : Common Language Runtime detected an invalid program."
```
        [Fact]
        public void Test1()
        {
            var source = """
                public class C 
                { 
                    static void Main()
                    {
                        var a = "a";
                        var b = "b";
                        var c = "c";

                        System.Console.WriteLine(System.String.Concat((System.ReadOnlySpan<char>)a, (System.ReadOnlySpan<char>)b) + c); 
                    }
                }
                """;

            var comp = CreateCompilation(source, options: TestOptions.ReleaseExe, targetFramework: TargetFramework.Net80);
            comp.MakeMemberMissing(WellKnownMember.System_String__op_Implicit_ToReadOnlySpanOfChar);
            comp.MakeMemberMissing(WellKnownMember.System_ReadOnlySpan_T__ctor_Reference);

            var verifier = CompileAndVerify(compilation: comp, expectedOutput: RuntimeUtilities.IsCoreClr8OrHigherRuntime ? "abc" : null, verify: /*ExecutionConditionUtil.IsCoreClr ? default : */Verification.Skipped).VerifyDiagnostics();

            //verifier.VerifyIL("C.Main", "");
        }
```